### PR TITLE
This commit updates container_loops.jl to use the new iteration protocol

### PR DESCRIPTION
**Pull request by StephenVavasis**
This commit rewrites container_loops.jl to incorporate the new iteration protocol
(introduced in 0.7.0-DEV) more thoroughly. It is not compatible with 0.6.

In addition, it fixes a bug in container_loops.jl in which the length()
function when applied to subranges (i.e., inclusive(a,b,c) or
exclusive(a,b,c)) returned the length of the whole container instead
of the length of the subrange. (There should be no value returned
for the length of the subrange since the data structure does not support
an O(1) algorithm or even O(log n) algorithm to compute the length.)

Some other smaller changes in this commit are as follows.

    IntSet was changed to BitSet (name change in 0.7.0-DEV)
    Small updates to documentation
    Some assert statements in balanced_tree.jl that were present
    during development are deleted.

### CODE REVIEW

**Comment by StephenVavasis** [comment]
I forgot to include in the commit message: this PR also rewrites `test_sorted_containers.jl` so that it runs much faster (10 sec versus 30+ sec).

File: src/balanced_tree.jl [hunk 1]
```diff
@@ -631,30 +631,30 @@ function compareInd(t::BalancedTree23, i1::Int, i2::Int)
...
     p2 = t.data[i2].parent
-    curdepth = t.depth
+    # curdepth = t.depth
     while true
-        @assert(curdepth > 0)
+        # @assert(curdepth > 0)
         if p1 == p2
             if i1a == t.tree[p1].child1
-                @assert(t.tree[p1].child2 == i2a || t.tree[p1].child3 == i2a)
+                # @assert(t.tree[p1].child2 == i2a || t.tree[p1].child3 == i2a)
```
Comment (kmsquire) [inline src/balanced_tree.jl]: Do you think you’ll ever uncomment these again?  If not, then it’s better to just delete them.

**Comment by StephenVavasis** [comment]
I won't ever uncomment those statements, but they do indicate invariants satisfied by the algorithm that may assist someone else in deciphering the code.  Is there a better way to indicate (with a comment) a program invariant?

File: src/balanced_tree.jl (hunk 1 above)
Comment (oxinabox) [reply src/balanced_tree.jl]: The curdepth commented out lines can be removed though. They are not asserts

**Comment by StephenVavasis** [comment]
Ok, I removed the commented-out asserts.  The `curdepth` variable in that section of the code existed only to support the `asserts`, so I deleted the commented-out statements involving `curdepth` as well.

**Comment by kmsquire** [comment]
I don’t know of a specific convention regarding invariants, but it would be nice to document them as such, to distinguish them from commented our code. Perhaps add `Invariant:` to the beginning of the line?

**Comment by oxinabox** [comment]
I am sorry, I think I have miscommunicated.

I would not have deleted them.
I think for now best would be to 
define a pair of macros:

`@invariant_assert(expr)`
and `@invariant_supporting(expr)`.

Which for now just effectively comment out their contents.
But later we can add some kind of mode that enables them.
A bit like how `@assert` works in other languages,
except very disabled even in debug mode.

The other reason not to delete them is while we may later wish to inform the compiler of these facts later for optimisation purposes.
(or if not these facts, similar facts in other places. The principle stands)
So we really do want to keep them documented.

**Comment by StephenVavasis** [comment]
OK, I backed out the last commit using `git reset` and changed the commented-out asserts to a new no-op macro as suggested.  However, I don't know how to upload this commit to github.  Below is the error message.  Git is suggesting that I first pull, but of course I don't want to pull; I just want this latest commit to supersede the previous one.  Any suggestions?  I'm not well-versed in git.

```
c:\Users\vavasis\OneDrive - University of Waterloo\GitHub2\DataStructures.jl\src>git push origin container_loop_fixes
To https://github.com/JuliaCollections/DataStructures.jl
 ! [rejected]        container_loop_fixes -> container_loop_fixes (non-fast-forward)
error: failed to push some refs to 'https://github.com/JuliaCollections/DataStructures.jl'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

**Comment by oxinabox** [comment]
I believe this calls for `git push --force`

**Comment by kmsquire** [comment]
As @oxinabox said, `git push --force` is the correct thing to do in this case.

**Comment by kmsquire** [comment]
Bump.  @StephanVavasis, is this good to go?  Do you need to force push?

**Comment by StephenVavasis** [comment]
Should be good to go... there are now no-op macros to indicate program invariants.

**Pull request merged**